src/iter.rs

```diff
@@ -113,3 +113,103 @@ where
     P: Write + Default,
 {
 }
+
+/// An iterator over the indices in a `TreeKey`.
+///
+/// The iterator yields `(indices: [usize; Y], depth: usize)`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct IndexIter<M: ?Sized, const Y: usize> {
+    /// Zero-size markers to allow being generic over M (by constraining the type parameters).
+    m: PhantomData<M>,
+
+    /// The iteration state.
+    ///
+    /// It contains the current field/element index at each path hierarchy level
+    /// and needs to be at least as large as the maximum path depth.
+    state: [usize; Y],
+
+    /// The remaining length of the iterator.
+    ///
+    /// It is used to provide an exact and trusted [Iterator::size_hint] ([core::iter::TrustedLen]).
+    ///
+    /// It may be None to indicate unknown length.
+    count: Option<usize>,
+}
+
+impl<M, const Y: usize> IndexIter<M, Y>
+where
+    M: TreeKey<Y> + ?Sized,
+{
+    pub(crate) fn new() -> Self {
+        let meta = M::metadata();
+        assert!(Y >= meta.max_depth);
+        let mut s = Self::new_unchecked();
+        s.count = Some(meta.count);
+        s
+    }
+
+    pub(crate) fn new_unchecked() -> Self {
+        Self {
+            count: None,
+            state: [0; Y],
+            m: PhantomData,
+        }
+    }
+}
+
+impl<M, const Y: usize> Iterator for IndexIter<M, Y>
+where
+    M: TreeKey<Y> + ?Sized,
+{
+    type Item = ([usize; Y], usize);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            return match M::traverse_by_key(self.state.iter().copied(), |_, _| Ok::<(), ()>(())) {
+                // Out of valid indices at the root: iteration done
+                Err(Error::NotFound(1)) => {
+                    debug_assert_eq!(self.count.unwrap_or_default(), 0);
+                    None
+                }
+                // Node not found at depth: reset current index, increment parent index,
+                // then retry path()
+                Err(Error::NotFound(depth @ 2..)) => {
+                    self.state[depth - 1] = 0;
+                    self.state[depth - 2] += 1;
+                    continue;
+                }
+                // Found a leaf at the root: leaf Option/newtype
+                // Since there is no way to end iteration by hoping for `NotFound` on a leaf Option,
+                // we force the count to Some(0) and trigger on that.
+                Ok(0) => {
+                    if self.count == Some(0) {
+                        None
+                    } else {
+                        debug_assert_eq!(self.count.unwrap_or(1), 1);
+                        self.count = Some(0);
+                        Some((self.state, 1))
+                    }
+                }
+                // Non-root leaf: advance index at current depth
+                Ok(depth) => {
```

Comment:
```suggestion
                Ok(depth @ 1..) => {
```

Maybe to future proof it a little better?